libs/designer-ui/src/lib/authentication/util.ts

```diff
@@ -432,3 +434,241 @@ export function containsToken(value: string): boolean {
+        }
+        errorMessage = checkForInvalidValues(parsedSerializedValue);
+        if (errorMessage) {
+          return errorMessage;
+        }
+      }
+    }
+  } catch {
+    return intl.formatMessage({
+      defaultMessage: 'Enter a valid JSON.',
+      description: 'Invalid JSON',
+    });
+  }
+
+  return '';
+};
+
+const authTypeConversion = (s: string): string => {
+  if (s === AuthenticationType.CERTIFICATE) {
+    return 'Client Certificate';
+  } else if (s === AuthenticationType.MSI) {
+    return 'Managed Service Identity';
+  } else if (s === AuthenticationType.OAUTH) {
+    return 'Active Directory OAuth';
+  }
+  return s;
+};
+
+/**
+ * Checks if any required property is missing.
+ * @arg {any} authentication -  The parsed authentication value.
+ * @arg {AuthenticationType} authType -  The authentication type.
+ * @return {string} - The error message for missing a required property.
+ */
+function checkForMissingOrInvalidProperties(authentication: any, authType: AuthenticationType): string {
+  const intl = getIntl();
+  let missingProperties: string[] = [];
+  const convertedAuthType = authTypeConversion(authType);
+  for (const key of PROPERTY_NAMES_FOR_AUTHENTICATION_TYPE[convertedAuthType]) {
+    if (key.isRequired && authentication[key.name] === undefined) {
+      missingProperties.push(key.name);
+    }
+  }
+  missingProperties = missingProperties.filter((name) => name !== AUTHENTICATION_PROPERTIES.MSI_IDENTITY.name);
+
+  if (authType === AuthenticationType.OAUTH) {
+    missingProperties = missingProperties.filter(
+      (name) =>
+        name !== AUTHENTICATION_PROPERTIES.AAD_OAUTH_CERTIFICATE_PFX.name &&
+        name !== AUTHENTICATION_PROPERTIES.AAD_OAUTH_CERTIFICATE_PASSWORD.name &&
+        name !== AUTHENTICATION_PROPERTIES.AAD_OAUTH_SECRET.name
+    );
+    if (missingProperties.length === 0) {
+      const authenticationKeys = Object.keys(authentication);
+      if (
+        (authenticationKeys.indexOf(AUTHENTICATION_PROPERTIES.AAD_OAUTH_CERTIFICATE_PASSWORD.name) !== -1 &&
+          authenticationKeys.indexOf(AUTHENTICATION_PROPERTIES.AAD_OAUTH_CERTIFICATE_PFX.name) !== -1 &&
+          authenticationKeys.indexOf(AUTHENTICATION_PROPERTIES.AAD_OAUTH_SECRET.name) !== -1) ||
+        (authenticationKeys.indexOf(AUTHENTICATION_PROPERTIES.AAD_OAUTH_CERTIFICATE_PASSWORD.name) === -1 &&
+          authenticationKeys.indexOf(AUTHENTICATION_PROPERTIES.AAD_OAUTH_CERTIFICATE_PFX.name) === -1 &&
+          authenticationKeys.indexOf(AUTHENTICATION_PROPERTIES.AAD_OAUTH_SECRET.name) === -1) ||
+        (authenticationKeys.indexOf(AUTHENTICATION_PROPERTIES.AAD_OAUTH_SECRET.name) !== -1 &&
+          authenticationKeys.indexOf(AUTHENTICATION_PROPERTIES.AAD_OAUTH_CERTIFICATE_PFX.name) !== -1) ||
+        (authenticationKeys.indexOf(AUTHENTICATION_PROPERTIES.AAD_OAUTH_SECRET.name) !== -1 &&
+          authenticationKeys.indexOf(AUTHENTICATION_PROPERTIES.AAD_OAUTH_CERTIFICATE_PASSWORD.name) !== -1)
```

Comment:
nit: Instead of `.indexOf() !== -1` you could just do `.includes()` here, no performance difference or anything, just a little more readable imo
Also this is a little hard to read, might help to just do the 3 checks and store them in descriptive variables to check over in 4 single-line groups

Reply:
yeah you're right, this code was just moved from the other code base, and I never refactored it. But will do that